File: Cpp/Lab2_7-2_8/Expression.cpp
#include "Expression.h"

#include <cmath>
#include "Exceptions.h"

double Expression::eval() {
    if(a*b+2 <= 0) throw LnNotGreaterThanZeroException();
    if(c == 0) throw DivisionByZeroException("c");
    double numerator = std::log(a*b+2)*c;
    double denominator = 41-b/c+1;
    if(denominator == 0) throw DivisionByZeroException("41-b/c+1");
    return numerator / denominator;
}

std::ostream &operator<<(std::ostream &os, const Expression &expr) {
    return os << "ln("<<expr.a<<" * "<<expr.b<<" + 2) * "<<expr.c
        <<"\n-----------------------\n"
        <<"41 - "<<expr.b<<"/"<<expr.c<<" + 1";
}

File: CSharp/Lab2_7-2_8/Lab2_7-2_8/Expression.cs
namespace Lab2_7_2_8;

public class Expression
{
    public double A { get; set; }
    public double B { get; set; }
    public double C { get; set; }

    public Expression(double a, double b, double c)
    {
        A = a;
        B = b;
        C = c;
    }
    public Expression() : this(1, 1, 1){}

    public double Eval()
    {
        if (A*B+2 <= 0) throw new LnNotGreaterThanZeroException();
        if (C == 0) throw new DivideByZeroException("Знаменник \"c\" дорівнює нулю!");
        double numerator = Math.Log(A*B+2)*C;
        double denominator = 41-B/C+1;
        if(denominator == 0) throw new DivideByZeroException("Знаменник \"41-b/c+1\" дорівнює нулю!");
        return numerator / denominator;
    }

    public override string ToString() => "ln("+A+" * "+B+" + 2) * "+C
                                         +"\n-----------------------\n"
                                         +"41 - "+B+"/"+C+" + 1";
}